flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java

```diff
@@ -90,24 +94,51 @@
 
         InputSelectable inputSelectable =
                 streamOperator instanceof InputSelectable ? (InputSelectable) streamOperator : null;
-        if (streamConfig.shouldSortInputs()) {
+
+        // this is a bit verbose because we're manually handling input1 and input2
+        // TODO: extract method
+        StreamConfig.InputConfig[] inputConfigs = streamConfig.getInputs(userClassloader);
+        boolean input1IsSorted = requiresSorting(inputConfigs[0]);
+        boolean input2IsSorted = requiresSorting(inputConfigs[1]);
+
+        if (input1IsSorted || input2IsSorted) {
+            // as soon as one input requires sorting we need to treat all inputs differently, to
+            // make sure that pass-through inputs have precedence
 
             if (inputSelectable != null) {
                 throw new IllegalStateException(
                         "The InputSelectable interface is not supported with sorting inputs");
             }
 
-            @SuppressWarnings("unchecked")
+            List<StreamTaskInput<?>> sortedTaskInputs = new ArrayList<>();
+            List<StreamTaskInput<?>> passThroughTaskInputs = new ArrayList<>();
+            int input1Index = 0;
```

Comment:
It is rather a matter of preference, but I'd move the index computing down below, where it is needed. I found it a bit confusing here and the IDE complains that in line 126 the `passThroughTaskInputs.size()` is always `0`.

We won't need this logic here, if we change the lines 151-160 to:
```StreamTaskInput<?>[] sortedInputs = selectableSortingInputs.getSortedInputs();
StreamTaskInput<?>[] passThroughInputs = selectableSortingInputs.getPassThroughInputs();
if (input1IsSorted) {
    input1 = toTypedInput(sortedInputs[0]);
} else {
    input1 = toTypedInput(passThroughInputs[0]);
}
if (input2IsSorted) {
    input2 = toTypedInput(sortedInputs[sortedInputs.length - 1]);
} else {
    input2 = toTypedInput(passThroughInputs[passThroughInputs.length - 1]);
}
```

We can base that on the assumption that the `input1` will be first in the corresponding group while the `input2` last.

Reply:
fixing